File: python/copass-core/src/copass_core/resources/sandbox_connections.py
"""Sandbox Connections resource — cross-user sandbox grants.

Port of ``typescript/packages/core/src/resources/sandbox-connections.ts``.

Phase 1A scope (read-only): ``list`` (backs the
``list_sandbox_connections`` management tool). Write methods
(``create``, ``revoke``, ``spawn_api_key``) are included for parity
with the TS surface — they're thin wrappers over existing endpoints.
Spec-driven write tools land in Phase 2.

Grants a non-owner (the *grantee*) ``viewer`` or ``editor`` access to
a sandbox the caller owns. Backed by ``sandbox_connections`` and
``copass_api_keys`` (migration 058).

Identity resolution: :meth:`create` accepts ``copass_id``, ``user_id``,
or ``email`` — exactly one. ``copass_id`` is resolved server-side to
the grantee's UUID; only the resolved UUID is persisted, so the grant
survives the grantee renaming or releasing their handle.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from copass_core.resources.base import BaseResource


_BASE = "/api/v1/storage/sandboxes"


class SandboxConnectionsResource(BaseResource):
    """``/api/v1/storage/sandboxes/{id}/connections``."""

    async def create(
        self,
        sandbox_id: str,
        *,
        role: str,
        copass_id: Optional[str] = None,
        user_id: Optional[str] = None,
        email: Optional[str] = None,
        project_id: Optional[str] = None,
        label: Optional[str] = None,
        expires_at: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Grant a connection on a sandbox you own. Owner-only."""
        body: Dict[str, Any] = {"role": role}
        if copass_id is not None:
            body["copass_id"] = copass_id
        if user_id is not None:
            body["user_id"] = user_id
        if email is not None:
            body["email"] = email
        if project_id is not None:
            body["project_id"] = project_id
        if label is not None:
            body["label"] = label
        if expires_at is not None:
            body["expires_at"] = expires_at
        return await self._post(f"{_BASE}/{sandbox_id}/connections", body)

    async def list(
        self,
        sandbox_id: str,
        *,
        include_revoked: bool = False,
    ) -> List[Dict[str, Any]]:
        """List all grants on a sandbox you own. Owner-only.

        Pass ``include_revoked=True`` to surface previously-revoked or
        expired grants too.
        """
        query: Dict[str, Any] = {}
        if include_revoked:
            query["include_revoked"] = "true"
        return await self._get(
            f"{_BASE}/{sandbox_id}/connections",
            query=query or None,
        )

    async def revoke(
        self,
        sandbox_id: str,
        connection_id: str,
    ) -> Dict[str, Any]:
        """Soft-delete a grant.

        Cascades to API keys bound to this connection so no key
        outlives its grant.
        """
        return await self._delete(
            f"{_BASE}/{sandbox_id}/connections/{connection_id}"
        )

    async def spawn_api_key(
        self,
        sandbox_id: str,
        connection_id: str,
    ) -> Dict[str, Any]:
        """Spawn a connection-scoped API key for an existing grant.

        The plaintext key is returned **exactly once** — persist it
        immediately or rotate.
        """
        return await self._post(
            f"{_BASE}/{sandbox_id}/connections/{connection_id}/api-keys"
        )


__all__ = ["SandboxConnectionsResource"]

File: typescript/packages/core/src/types/sandbox-connections.ts
/**
 * Cross-user sandbox grants — POST/GET /sandboxes/{sandbox_id}/connections,
 * DELETE /{connection_id}, POST /{connection_id}/api-keys.
 *
 * Connections let a sandbox owner grant a teammate access to their
 * sandbox. The grantee gets a connection-scoped API key and acts
 * inside the owner's sandbox at `viewer` or `editor` role.
 */

export type ConnectionRole = 'viewer' | 'editor';
export type ConnectionStatus = 'active' | 'revoked' | 'expired';

export interface SandboxConnection {
  connection_id: string;
  sandbox_id: string;
  /** Grantee's platform UUID — NOT the sandbox owner. */
  user_id: string;
  /**
   * Grantee's *current* Copass handle (without leading `@`), resolved
   * at read time. Absent if the grantee has not claimed a handle.
   * The historical handle (when the grant was minted by `copass_id`)
   * is preserved in `label` instead.
   */
  copass_id?: string | null;
  role: ConnectionRole | string;
  /** Optional project scope. NULL means the grant covers the whole sandbox. */
  project_id?: string | null;
  label?: string | null;
  expires_at?: string | null;
  revoked_at?: string | null;
  created_by: string;
  created_at: string;
  status: ConnectionStatus | string;
}

export interface CreateSandboxConnectionRequest {
  /**
   * Grantee Copass handle (e.g. `arcus` or `@arcus`). Resolved to a
   * `user_id` at create-time; only the resolved UUID is persisted.
   * Mutually exclusive with `user_id` and `email`.
   */
  copass_id?: string;
  /** Grantee platform UUID. Mutually exclusive with `copass_id` and `email`. */
  user_id?: string;
  /**
   * Grantee email. Reserved for a future invite flow; the server
   * returns 501 today if this is the only identifier supplied. Use
   * `copass_id` or `user_id` instead.
   */
  email?: string;
  role: ConnectionRole;
  /** Optional project-level scope inside this sandbox. */
  project_id?: string;
  label?: string;
  /** ISO 8601 timestamp. NULL = never expires until revoked. */
  expires_at?: string;
}

export interface ListSandboxConnectionsOptions {
  /** When true, includes revoked grants. Defaults to false. */
  include_revoked?: boolean;
}

export interface CreateSandboxConnectionApiKeyResponse {
  api_key_id: string;
  /**
   * Full plaintext API key (`olk_…`). Returned exactly once at
   * creation; the server does not retain plaintext after this
   * response. Persist it immediately or rotate.
   */
  plaintext_key: string;
  key_prefix: string;
  expires_at?: string | null;
  warning?: string;
}
